ouroboros-consensus/test/consensus-test/Test/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -312,99 +412,164 @@ runChainSync securityParam (ClientUpdates clientUpdates)
+    let checkTipTime :: m ()
+        checkTipTime = do
+            (now, candidates) <- do
+              now        <- systemTimeCurrent systemTime
+              candidates <-
+                atomically $ readTVar varCandidates >>= traverse readTVar
+              pure (now, candidates)
+            forM_ candidates $ \candidate -> do
+              let p = castPoint $ AF.headPoint candidate :: Point TestBlock
+              case pointSlot p of
+                Origin  -> pure ()
+                At slot -> when (now < toOnset slot) $ do
+                  atomically $ writeTVar varClientResult $ Just
+                    $ ClientSelectedFutureTip $ FutureTip {
+                          ftNow   = now
+                        , ftPoint = (toOnset slot, p)
+                        }
+
+    do
+      let loop tick = do
+              -- first update the clocks
+              advanceWallClockForTick tick
+              atomically $ writeTVar varCurrentLogicalTick tick
+
+              -- then do the messages
+              updateChainsDuringTick tick
+              when (tick == startSyncingAt) $ initiateChainSync
+
+              -- check the invariants before advancing the clock again
+              --
+              -- This is not a perfect check, since the server's chain may have
+              -- violated the invariant ephemerally (ie due to a subsequent
+              -- rollback during the same logical tick). However, other
+              -- QuickCheck seeds/counterexamples should trigger such a bug in
+              -- a non-ephemeral way.
+              checkTipTime
+
+              when (tick < finalTick) $ loop (tick + 1)
+      loop (Tick 1)
+
+      -- This delay seems enough to let all threads finish their final work.
+      --
+      -- TODO what is the necessary threshold?
+      threadDelay 86400
 
     traceEvents <- getTrace
     -- Collect the return values
     atomically $ do
-      finalClientChain       <- readTVar varClientState
-      finalServerChain       <- chainState <$> readTVar varChainProducerState
+      finalClientChain  <- readTVar varClientState
+      finalServerChain  <- chainState <$> readTVar varChainProducerState
       candidateFragment <- readTVar varFinalCandidates >>= readTVar . (Map.! serverId)
-      mbResult      <- readTVar varClientResult
+      mbResult          <- readTVar varClientResult
       return ChainSyncOutcome {
-        syncedFragment   = AF.mapAnchoredFragment testHeader candidateFragment
-        , ..
+          finalClientChain
+        , finalServerChain
+        , mbResult
+        , syncedFragment   = AF.mapAnchoredFragment testHeader candidateFragment
+        , traceEvents
         }
   where
     k = maxRollbacks securityParam
 
-    slotLength :: SlotLength
-    slotLength = slotLengthFromSec 20
+    toOnset :: SlotNo -> RelativeTime
+    toOnset slot = RelativeTime $
+        scaleNominalDelay
+          (getSlotLength slotLength)
+          (unSlotNo slot)
```

Comment:
What about making this a top-level declaration (instead of `scaleNominalDelay), such that it can also be used in `tickArrivalTimes`?